datajoint/external.py

```diff
@@ -0,0 +1,109 @@
+import os
+from . import config, DataJointError
+from .hash import long_hash
+from .blob import pack, unpack
+from .base_relation import BaseRelation
+from .declare import STORE_HASH_LENGTH, HASH_DATA_TYPE
+
+
+class ExternalTable(BaseRelation):
+    """
+    The table tracking externally stored objects.
+    Declare as ExternalTable(connection, database)
+    """
+    def __init__(self, arg, database=None):
+        if isinstance(arg, ExternalTable):
+            super().__init__(arg)
+            # copy constructor
+            self.database = arg.database
+            self._connection = arg._connection
+            return
+        super().__init__()
+        self.database = database
+        self._connection = arg
+        if not self.is_declared:
+            self.declare()
+
+    @property
+    def definition(self):
+        return """
+        # external storage tracking 
+        hash  : {hash_data_type}  # the hash of stored object + store name
+        ---
+        size      :bigint unsigned   # size of object in bytes
+        timestamp=CURRENT_TIMESTAMP  :timestamp   # automatic timestamp
+        """.format(hash_data_type=HASH_DATA_TYPE)
+
+    @property
+    def table_name(self):
+        return '~external'
+
+    def put(self, store, obj):
+        """
+        put an object in external store
+        """
+        try:
+            spec = config[store]
+        except KeyError:
+            raise DataJointError('Storage {store} is not configured'.format(store=store))
+
+        # serialize object
+        blob = pack(obj)
+        hash = long_hash(blob) + store[len('external-'):]
```

Comment:
```store[len('external-'):]``` seems incongruent w/r/t line 88: ```store = 'external' + ('-' if store else '') + store``` which could conceivably allow only ```external```; that said using only 'external' as a specifier is already blocked in  ```declare.py``` so I'm not sure how important this is.. perhaps line 88 should throw an error rather than allowing 'external' only?

Reply:
The idea is to allow either `external` or `external-somethin` where `somethin` cannot exceed 8 characters.  `external` by itself should not be blocked.  If it is, it's my mistake and I will fix it and add a test for it.  Many people will only have one external storage and they will just use `external`.

Reply:
Why is `store[len('external-'):]` incongruent?  You can do `a[n:]` where `n` exceeds the size of `a` no problem.

Reply:
Ah ok - I think on first read I was taking this to mean there was an explicit expectation that the slice would return something to append to the hash, which was then "confirmed" when I saw that declare.py requires the external-<something> convention (via the 'isidentifier' check on the secondary portion) and throws exceptions suggesting this is a requirement.

Not allowing the bare 'external' in some cases (here, declare.py) but allowing it in another (line 88) is why I proposed one was incongruent with respect to the other.

Reply:
if 'external' alone is is allowed, it looks like the store_name parts from declare (line 191+) need to be altered -
```
>>> storestr='external'
>>> storestr.split('-')
['external']
>>> store_name = storestr.split('-')
>>> store_name[1:]
[]
>>> ''.join(store_name[1:])
''
>>> ''.join(store_name[1:]).isidentifier()
False
```